README.md
# oddb2xml

[![Build Status](https://github.com/zdavatz/oddb2xml/workflows/Ruby/badge.svg)](https://github.com/zdavatz/oddb2xml/actions)

* oddb2xml -a nonpharma -o fi

creates the following xml files:

* oddb_substance.xml
* oddb_limitation.xml
* oddb_interaction.xml
* oddb_code.xml
* oddb_product.xml
* oddb_article.xml
* oddb_fi.xml
* oddb_fi_product.xml

and

* oddb2xml -f dat
* oddb2xml -f dat -a nonpharma

creates .dat files according to ([IGM-11](http://dev.ywesee.com/uploads/att/IGM.pdf)). IGM-11 describes the structure of the zurrose_transfer.dat.

* oddb.dat
* oddb_with_migel.dat

the files are using [swissINDEX](http://www.refdata.ch/downloads/company/download/swissindex_TechnischeBeschreibung.pdf), [BAG-XML](http://bag.e-mediat.net/SL2007.Web.External/Default.aspx?webgrab=ignore) and [Swissmedic](http://www.swissmedic.ch/daten/00080/00251/index.html?lang=de) as sources.

The following additional data is in the files:

* [Wirkstoffe](http://bag.e-mediat.net/SL2007.Web.External/File.axd?file=XMLPublications.zip) (BAG XML)
* [Kühlkette](http://www.swissmedic.ch/daten/00080/00254/index.html?lang=de&download=NHzLpZeg7t,lnp6I0NTU042l2Z6ln1acy4Zn4Z2qZpnO2Yuq2Z6gpJCDdH57e2ym162epYbg2c_JjKbNoKOn6A--) (Swissmedic)
* [Orphan Drugs](http://www.swissmedic.ch/daten/00081/index.html?lang=de&download=NHzLpZeg7t,lnp6I0NTU042l2Z6ln1acy4Zn4Z2qZpnO2Yuq2Z6gpJCDdH55f2ym162epYbg2c_JjKbNoKSn6A--&.xls) (Swissmedic)
* [FI: de, fr](http://download.swissmedicinfo.ch) (Swissmedic)
* Limitation-Texte (BAG XML)
* Interaktionen [EPha.ch](http://epha.ch)
* [Betäubungsmittel](http://www.swissmedic.ch/produktbereiche/00447/00536/index.html?lang=de&download=NHzLpZeg7t,lnp6I0NTU042l2Z6ln1acy4Zn4Z2qZpnO2Yuq2Z6gpJCDdH1,fWym162epYbg2c_JjKbNoKSn6A--&.pdf) und psychotrope Stoffe (Swissmedic)
* Non-Pharma from Refdata and Suppliers (swissINDEX)

The top elements of all XML files have a SHA256 attribute over their content. The content corresponds to Nokogiris text method of the node which is essentially join by "\n" + some whitespaces of each element.
Consumers of the data file may use it to check whether they have to replace the corresponding nodes.

Generating files for Elexis Artikelstamm is discussed in the [Readme for the Artikelstamm](artikelstamm.md)


## usage

HIN (http://hin.ch) creates daily the actual file. They can be downloaded from `https://download.hin.ch/download/oddb2xml`, e.g. using  `wget https://download.hin.ch/download/oddb2xml/oddb_article.xml`

see `--help`.

```
    /opt/src/oddb2xml_v5/bin/oddb2xml version 2.4.3
    Usage:
    oddb2xml [option]
      produced files are found under data
    -a, --append              Additional target nonpharma
    -r, --artikelstamm        Create Artikelstamm Version 3 and 5 for Elexis >= 3.1
    -c, --compress-ext=<s>    format F. {tar.gz|zip}
    -e, --extended            pharma, non-pharma plus prices and non-pharma from zurrose.
                                                          Products without EAN-Code will also be listed.
                                                          File oddb_calc.xml will also be generated
    -f, --format=<s>          File format F, default is xml. {xml|dat}
                                                          If F is given, -o option is ignored. (Default: xml)
    -i, --include             Include target option for ean14  for 'dat' format.
                                                          'xml' format includes always ean14 records.
    -I, --increment=<i>       Increment price by x percent. Forces -f dat -p zurrose.
                                                          create additional field price_resellerpub as
                                                          price_extfactory incremented by x percent (rounded to the next 0.05 francs)
                                                          in oddb_article.xml. In generated zurrose_transfer.dat PRPU is set to this price
                                                          Forces -f dat -p zurrose.
    -o, --fi                  Optional fachinfo output.
    -p, --price               Price source (transfer.dat) from ZurRose
    -t, --tag-suffix=<s>      XML tag suffix S. Default is none. [A-z0-9]
                                                          If S is given, it is also used as prefix of filename.
    -x, --context=<s>         {product|address}. product is default. (Default: product)
    -l, --calc                create only oddb_calc.xml with GTIN, name and galenic information
    -s, --skip-download       skips downloading files it the file is already under downloads.
                                                          Downloaded files are saved under downloads
    --log                     log important actions
    -u, --use-ra11zip=<s>     Use the ra11.zip (a zipped transfer.dat from Galexis)
    -v, --version             Print version and exit
    -h, --help                Show this message
```

## Option examples

```
$ oddb2xml -t md                        # => md_article.xml, md_product.xml, md_substance.xml
$ oddb2xml -a nonpharma -t md -c tar.gz # => md_xml_dd.mm.yyyy_hh.mm.tar.gz
$ oddb2xml -f dat                       # => oddb.dat
$ oddb2xml -f dat -a nonpharma          # => oddb_with_migel.dat
$ oddb2xml -e                           # => oddb_article.xml
```

output.

```
$ oddb2xml
DE
        Pharma products: 14801
FR
        Pharma products: 14801
```

## Supported ruby version

We run tests on github for the Ruby versions mentioned in the .github/workflows/ruby.ymlfile. You will need ruby >= 3.1.0 to work correctly.
Look at the github actions to see the spec test results. We ignore errors against ruby head, as it is sometimes unstable.


## XSD files

The file oddb2xml.xsd was manually created by merging the output of the xmlbeans tools inst2xsd and trang

* http://xmlbeans.apache.org/docs/2.0.0/guide/tools.html#inst2xsd
* http://www.thaiopensource.com/relaxng/trang.html

Running rake spec will validated the XML-files generated during the tests using the Nokogiri validator.
We have two XSD files. One for oddb_calc.xml and one for the rest.

Manually you can also validate (assuming that you have installed the xmlbeans tools) all generated XML-files using

* xsdvalidate oddb_calc.xsd oddb_article.xml oddb_calc.xml
* xsdvalidate oddb2xml.xsd oddb_article.xml oddb_code.xml oddb_interaction.xml oddb_product.xml oddb_substance.xml

## XML files

xml files generated are:

* oddb_substance.xml
* oddb_limitation.xml
* oddb_interaction.xml
* oddb_code.xml
* oddb_product.xml
* oddb_article.xml
* oddb_fi.xml
* oddb_fi_product.xml

### article.xml

oddb2xml creates article.xml as oddb_article.xml by default.

```
<?xml version="1.0" encoding="utf-8"?>
<ARTICLE xmlns:xsd="http://www.w3.org/2001/XMLSchema" xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance" xmlns="http://wiki.oddb.org/wiki.php?pagename=Swissmedic.Datendeklaration" CREATION_DATETIME="2015-09-09T09:50:28+0000" PROD_DATE="2015-09-09T09:50:28+0000" VALID_DATE="2015-09-09T09:50:28+0000">
  <ART DT="2015-09-09 00:00:00 +0000" SHA256="896dd24bfb4cfd56dcfd3709150da9b652626a430adefbe57cb405a9d46684c6">
    <REF_DATA>1</REF_DATA>
    <PHAR>2731179</PHAR>
    <SMCAT>D</SMCAT>
    <SMNO>16105058</SMNO>
    <PRODNO>161051</PRODNO>
    <VAT>2</VAT>
    <SALECD>A</SALECD>
    <CDBG>N</CDBG>
    <BG>N</BG>
    <DSCRD>HIRUDOID Creme 3 mg/g 40 g</DSCRD>
    <DSCRF>HIRUDOID crème 3 mg/g 40 g</DSCRF>
    <SORTD>HIRUDOID CREME 3 MG/G 40 G</SORTD>
    <SORTF>HIRUDOID CRèME 3 MG/G 40 G</SORTF>
    <SYN1D>Hirudoid</SYN1D>
    <SYN1F>Hirudoid</SYN1F>
    <SLOPLUS>2</SLOPLUS>
    <ARTCOMP>
      <COMPNO>7601001002258</COMPNO>
    </ARTCOMP>
    <ARTBAR>
      <CDTYP>E13</CDTYP>
      <BC>7680161050583</BC>
      <BCSTAT>A</BCSTAT>
    </ARTBAR>
    <ARTPRI>
      <PTYP>PEXF</PTYP>
      <PRICE>4.768575</PRICE>
    </ARTPRI>
    <ARTPRI>
      <PTYP>PPUB</PTYP>
      <PRICE>8.8</PRICE>
    </ARTPRI>
    <ARTPRI>
      <PTYP>ZURROSE</PTYP>
      <PRICE>4.77</PRICE>
    </ARTPRI>
    <ARTPRI>
      <PTYP>ZURROSEPUB</PTYP>
      <PRICE>8.80</PRICE>
    </ARTPRI>
    <ARTINS>
      <NINCD>10</NINCD>
    </ARTINS>
  </ART>
  ...
  <RESULT>
    <OK_ERROR>OK</OK_ERROR>
    <NBR_RECORD>14801</NBR_RECORD>
    <ERROR_CODE/>
    <MESSAGE/>
  </RESULT>
</ARTICLE>
```

### product.xml

For example, if `-t _swiss` is given then oddb2xml creates product.xml as swiss_product.xml.

```
<?xml version="1.0" encoding="utf-8"?>
<PRODUCT_SWISS xmlns:xsd="http://www.w3.org/2001/XMLSchema" xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance" xmlns="http://wiki.oddb.org/wiki.php?pagename=Swissmedic.Datendeklaration" CREATION_DATETIME="2012-11-21T13:01:29.5903756+0900" PROD_DATE="2012-11-21T13:01:29.5903756+0900" VALID_DATE="2012-11-21T13:01:29.5903756+0900">
  <PRD_SWISS DT="" SHA256="aa82eee2d542787cf2cb8b7f17d748223ec723b935ce20cd29d89e284d16fea1">
    <GTIN>7680353660163</GTIN>
    <PRODNO>353661</PRODNO>
    <DSCRD>KENDURAL Depottabl 30 Stk</DSCRD>
    <DSCRF>KENDURAL cpr dépot 30 pce</DSCRF>
    <ATC>B03AE10</ATC>
    <IT>06.07.1.</IT>
    <CPT>
      <CPTCMP>
        <LINE>0</LINE>
        <SUBNO>5</SUBNO>
        <QTY>105</QTY>
        <QTYU>mg</QTYU>
      </CPTCMP>
      <CPTCMP>
        <LINE>1</LINE>
        <SUBNO>1</SUBNO>
        <QTY>500</QTY>
        <QTYU>mg</QTYU>
      </CPTCMP>
    </CPT>
    <PackGrSwissmedic>30</PackGrSwissmedic>
    <EinheitSwissmedic>Tablette(n)</EinheitSwissmedic>
    <SubstanceSwissmedic>ferrum(II), acidum ascorbicum</SubstanceSwissmedic>
    <CompositionSwissmedic>ferrum(II) 105 mg ut ferrosi sulfas dessiccatus, acidum ascorbicum 500 mg ut natrii ascorbas, color.: E 124, excipiens pro compresso obducto.</CompositionSwissmedic>
  </PRD>
  ...
  <RESULT_SWISS>
    <OK_ERROR_SWISS>OK</OK_ERROR_SWISS>
    <NBR_RECORD_SWISS>14336</NBR_RECORD_SWISS>
    <ERROR_CODE_SWISS/>
    <MESSAGE_SWISS/>
  </RESULT_SWISS>
</PRODUCT_SWISS>
```

### substance.xml

product.xml has relation to substance as `<SUBNO>`.

```
<?xml version="1.0" encoding="utf-8"?>
<SUBSTANCE xmlns:xsd="http://www.w3.org/2001/XMLSchema" xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance" xmlns="http://wiki.oddb.org/wiki.php?pagename=Swissmedic.Datendeklaration" CREATION_DATETIME="2012-12-11T14:27:17.4444763+0900" PROD_DATE="2012-12-11T14:27:17.4444763+0900" VALID_DATE="2012-12-11T14:27:17.4444763+0900">
  <SB DT="" SHA256="a510f9b1e7216cda2d5e0c3b82bacef96da963e14f36c97e0e1a8baf55d00287">
    <SUBNO>1</SUBNO>
    <NAML>Acidum ascorbicum (Vitamin C, E300)</NAML>
  </SB>
  <SB DT="" SHA256="de64fcc718b7f30bfe4283fb40c8b558cf2f30a8acc4a7bf6a643e82dfe82931">
    <SUBNO>2</SUBNO>
    <NAML>Alprostadilum</NAML>
  </SB>
  ...
  <RESULT>
    <OK_ERROR>OK</OK_ERROR>
    <NBR_RECORD>1441</NBR_RECORD>
    <ERROR_CODE/>
    <MESSAGE/>
  </RESULT>
</SUBSTANCE>
```

## Data sources

We use the following files:

* https://www.swissmedic.ch/arzneimittel/00156/00221/00222/00230/index.html?lang=de (Präparateliste und zugelassene Packungen)
* https://raw.githubusercontent.com/zdavatz/oddb2xml_files/master/interactions_de_utf8.csv
* http://refdatabase.refdata.ch/Service/Article.asmx
* http://bag.e-mediat.net/SL2007.Web.External/File.axd?file=XMLPublications.zip
* https://www.medregbm.admin.ch/Publikation/CreateExcelListBetriebs
* https://www.medregbm.admin.ch/Publikation/CreateExcelListMedizinalPersons
* http://zurrose.com/fileadmin/main/lib/download.php?file=/fileadmin/user_upload/downloads/ProduktUpdate/IGM11_mit_MwSt/Vollstamm/transfer.dat
* https://index.ws.e-mediat.net/Swissindex/NonPharma/ws_NonPharma_V101.asmx
* https://index.ws.e-mediat.net/Swissindex/NonPharma/ws_Pharma_V101.asmx
* http://download.swissmedicinfo.ch/ (AipsDownload)
* https://raw.githubusercontent.com/zdavatz/oddb2xml_files/master/LPPV.txt
* https://raw.githubusercontent.com/epha/robot/master/data/manual/swissmedic/atc.csv

## Rules for matching GTIN (aka EAN13), product number and IKSNR

For drugs which appear in Packungen.xlsx file published by Swissmedic the following rule is used to create the GTIN
* First 4 digits identify SwissMedic and are fixed to 7680
* next 5 digits corresponding to IKSNR (authorization) number
* next 3 digits corresponding to Packungscode
* last digit is checksum

The product number is calculated as
* 5 digits corresponding to IKSNR (authorization) number
* 2 digits corresponding to Dosisstärke (aka sequence number)

In oddb_article.xml you find
* GTIN is found as "BC" inside "ARTBAR"
* The product number as field PRODNO

Example given. For the IKSNR 48305 sequence number 1 named "Felden, Gel" with Packungscode "024" we get GTIN 7680483050247 and a product number 483051.


## SSLv3 cert for Windows Users

Some websites need SSLv3 connection.
If you don't have these root CA files (x509), Please install these Certificates before running.  
see [cURL Website](http://curl.haxx.se/ca/)

You can confirm wit `ruby -ropenssl -e 'p OpenSSL::X509::DEFAULT_CERT_FILE'`.

### Windows User: Making your SSL Certificate permanent via your PATH

1. Download this [cacert.pem](http://curl.haxx.se/ca/cacert.pem) (cURL) into your HOME directory.
 * or directly select cacert.pem from your oddb2xml-x.x.x gems directory.
 * tools/cacert.pem is bundled with the oddb2xml gem.

History.txt
=== 2.7.10 / 2.10.23  -- not yet released
* Fix spec tests for ruby 3.2
* Removed ruby 2.6 from supported versions
* Remove unused italian translation. Elexis never adopted it.

=== 2.7.9 / 19.09.22
* Remove newly generated DSCRI when not running --artikelstamm and 
  generate always (as before 2.7.8) a DSCRF fiele

=== 2.7.8 / 14.09.22
* Add fields for french and italian in artikelstamm whenever possible

=== 2.7.7/ 12.08.2022
* Workaround for artikelstamm V3. Strip IKSCAT from A+ to A to be compatible with XSD

=== 2.7.6/ 9.03.2022
* Workaround for artikelstamm V5. Strip IKSCAT from A+ to A to be compatible with XSD
* Fix spec tests

=== 2.7.5 / 1.02.2022
* Specify old versions of savon and rdoc to avoid importing psych
* Added ruby 3.1 to tested ruby versions
* Removed ruby 2.5 from supported versions
* Force psych to < 4.0 to avoid the error Psych::DisallowedClass: Tried to load unspecified class: Struct::GalenicGroup
*   Therefor rdoc < 6.4

=== 2.7.4 / 30.03.2021
* Update needed Ruby-Version to 2.5.0 in gemspec
* Ensure downloads directory is created at first run

=== 2.7.3 / 29.03.2021
* Do not delete swissmedic_package.xlsx and swissmedic_orphan.xlsx
* Save swissmedic_package.xlsx and swissmedic_orphan.xlsx under downloads
* Log removing files

=== 2.7.2 / 18.03.2021
* Added https://github.com/testdouble/standard and fixed all useful warnings
* Added standard to rake tasks
* Use ATC-code from refdata if available and show differences
* Get pharmacode from tranfer.dat for artikelstamm if possible
* We do not longer test Ruby 2.4 with github/actions
* Update build status to use github/actions instead of travis

=== 2.7.1 / 03.02.2021
* travis-ci -> Github Actions
* limit DSCRD/DSCRF to 120 chars as specified in the XSD
* Use huge option for nokogiri as suggested by  https://github.com/sparklemotion/nokogiri/issues/2187
* Adapt artikelstamm to refdata without pharmacode
* oddb_article get pharmacode where possible from transfer.dat

=== 2.7.0 / 05.11.2020
* Chapter70 hack nur als Kommentar
* Avoid double hyphen in comment

=== 2.6.9 / 29.06.2020

* Fix problem when running bin/oddb2xml -a -f dat. Thanks to Lukas Furre for reporting the problem
* Fix warning with regular expression thas has ']'

=== 2.6.8 / 26.06.2020

* Re-Add support for Ruby Version 2.4
* Specified minimal ruby version in gemspec
* Patch artikelstamm build to emit a product for Varilrix

=== 2.6.7 / 25.06.2020
* Added value 'C' (Co-marketing) for field GENERIC_TYPE Elexis_Artikelstamm_v5.xsd from BAG Preparations.xml
* Fixed limitations coming from the chapte 70 hack
* Ruby 2.7 compatible (No URI.open warning)

=== 2.6.6 / 11.01.2020
* Fix link for varia_De.htm

=== 2.6.5 / 11.01.2020
* Fix issue #37: Update BAG Download Link
** Pending problem: varia_De.htm for Artikelstamm cannot be found anymore

=== 2.6.4 / 14.08.2019
* Fix issue #35: Missing product with ATC code for 7680005850010 Varilrix when building Artikelstamm_V5

=== 2.6.3 / 01.07.2019
* Use https://github.com/zdavatz/cpp2sqlite/blob/master/input/atc_codes_multi_lingual.txt instead of http://download.epha.ch/data/atc/atc.csv

=== 2.6.2 / 01.07.2019
* Use https://raw.githubusercontent.com/zdavatz/oddb2xml_files/master/interactions_de_utf8.csv instead of https://download.epha.ch/cleaned/matrix.csv

=== 2.6.1 / 27.02.2019
* Update rubyXL to 3.4.0 with fix to avoid problem reported via https://github.com/weshatheleopard/rubyXL/issues/325
* When generating artikelstamm XML generate products (aka sequences) with name like "3TC Filmtabl 150 mg" instead of only "3TC"
  This corrects an error introduced by version 2.5.9
* Adapt to changed link description "Zugelassene Packungen" in https://www.swissmedic.ch/swissmedic/de/home/services/listen_neu.html

=== 2.6.0 / 13.02.2019
* Adapt to changed xlsx files from swissmedic

=== 2.5.9 / 06.12.2018

* Only changes when generating artikelstamm
** Use sequence name from packages when there is no german name in the article (from refdata)
** Correct header of artikelstamm-*.csv

=== 2.5.8 / 20.09.2018

* Set SALECD to 'A' if article is from refdata_pharma/swissmedic
  Marking these cases in oddb_article.xml, eg.
  <SALECD>A<!--Overriding status I nincd 10 for 7680669940041 as in refdata_pharma--></SALECD>

=== 2.5.7 / 11.09.2018

* Emit ATC code if product is not in refdata

=== 2.5.6 / 21.08.2018

* Artikelstamm: Add homeopathic drugs of chapter 70 http://bag.e-mediat.net/Sl2007.web.external/varia_De.htm
* Artikelstamm: Replace HTML entities by UTF-8 characters for limitations

=== 2.5.5 / 09.05.2018

* Fix parsing 66540 Tektrotyd, Markierungsbesteck
* Update QA.md/artikelstamm.md

=== 2.5.4 / 09.04.2018

* Correct name in Artikelstamm, when several packages exist in Preparions.xml (SL)
* Mark correctly all SL-Entries in Artikeltstamm

=== 2.5.3 / 28.03.2018

* Updated needed Ruby version in README.md
* Test using Ruby 2.5.0 on tracis-ci, too
* Fix spec tests on travis-ci
* Added bin/check_artikelstamm
* Fixed problem with prodno in Artikeltstamm

=== 2.5.2 / 21.03.2018

*  Artikelstamm
** Add prices for GTIN with 14 chars
** Add ATC-Code for products like priorix
* Handle GTIN with size of 8,12,14 correctly
* Don't call exit(2) if xmllint is not present
* Corrected formatting of artikelstamm.md

=== 2.5.1 / 09.02.2018

* Updated requirement for nokogiri to >= 1.8.2
* Artikelstamm_V5 seems to work with the following additions
** Create Elexis_Artikelstamm_v5.csv with most used fields, including SL
** Better reporting of downloads
** Allow overriding a specific field of an item
** trace data_origin
** Better spec tests for options
** Trollop instead of optparse

=== 2.5.0 / 06.02.2018

* Adapt to swissmedic fixing a typo in Packungen.XLSX header row

=== 2.4.9 / 15.01.2018

* Adapt to new link of epha atc.csv
* Ensure that transfer.dat lines have always the correct line length of 97

=== 2.4.8 / 20.11.2017

* Adapt to new links of swissmedic download files

=== 2.4.7 / 26.10.2017

* Only emit PHAR if present


lib/oddb2xml/downloader.rb
# encoding: utf-8

require 'net/ntlm/version' # needed to avoid error: uninitialized constant Net::NTLM::VERSION
require 'rubyntlm'
require 'mechanize'
require 'zip'
require 'savon'

SkipMigelDownloader = true  # https://github.com/zdavatz/oddb2xml_files/raw/master/NON-Pharma.xls

module Oddb2xml
  module DownloadMethod
    private
    def download_as(file, option='r')
      tempFile  = File.join(WorkDir,   File.basename(file))
      file2save = File.join(Downloads, File.basename(file))
      report_download(@url, file2save)
      data = nil
      FileUtils.rm_f(tempFile, :verbose => false)
      if Oddb2xml.skip_download(file)
        io = File.open(file, option)
        data = io.read
      else
        begin
          response = @agent.get(@url)
          response.save_as(file)
          response = nil # win
          io = File.open(file, option)
          data = io.read
        rescue Timeout::Error, Errno::ETIMEDOUT
          retrievable? ? retry : raise
        ensure
          io.close if io and !io.closed? # win
          Oddb2xml.download_finished(tempFile)
        end
      end
      return data
    end
  end
  class Downloader
    attr_reader :type, :agent
    def initialize(options={}, url=nil)
      @options     = options
      @url         = url
      @retry_times = 3
      HTTPI.log = false # disable httpi warning
      Oddb2xml.log "Downloader from #{@url} for #{self.class}"
      init
    end
    def report_download(url, file)
      Oddb2xml.log sprintf("%-20s: download_as %-24s from %s",
                           self.class.to_s.split('::').last,
                           File.basename(file),
                           url)
    end
    def init
      @agent = Mechanize.new
      @agent.user_agent = 'Mozilla/5.0 (X11; Linux x86_64; rv:16.0) Gecko/20100101 Firefox/16.0'
      @agent.redirect_ok         = true
      @agent.redirection_limit   = 5
      @agent.follow_meta_refresh = true
      if RUBY_PLATFORM =~ /mswin|mingw|bccwin|cygwin/i and
         ENV['SSL_CERT_FILE'].nil?
        cert_store = OpenSSL::X509::Store.new
        cert_store.add_file(File.expand_path('../../../tools/cacert.pem', __FILE__))
        @agent.cert_store = cert_store
      end
      @agent.verify_mode = OpenSSL::SSL::VERIFY_NONE
    end
    protected
    def retrievable?
      if @retry_times > 0
        sleep 5
        @retry_times -= 1
        true
      else
        false
      end
    end
    def read_xml_from_zip(target, zipfile)
      Oddb2xml.log "read_xml_from_zip target is #{target} zip: #{zipfile} #{File.exists?(zipfile)}"
      if Oddb2xml.skip_download?
        entry = nil
        Dir.glob(File.join(Downloads, '*')).each { |name| if target.match(name) then entry = name; break end }
        if entry
          dest = "#{Downloads}/#{File.basename(entry)}"
          if File.exists?(dest)
            Oddb2xml.log "read_xml_from_zip return content of #{dest} #{File.size(dest)} bytes "
            return IO.read(dest)
          else
            Oddb2xml.log "read_xml_from_zip could not read #{dest}"
          end
        else
          Oddb2xml.log "read_xml_from_zip could not find #{target.to_s}"
        end
      end
      xml = ''
      if RUBY_PLATFORM =~ /mswin|mingw|bccwin|cygwin/i
        Zip::File.open(zipfile) do |zipFile|
          zipFile.each do |entry|
            if entry.name =~ target
              Oddb2xml.log "read_xml_from_zip reading #{__LINE__}: #{entry.name}"
              io = entry.get_input_stream
              until io.eof?
                bytes = io.read(1024)
                xml << bytes
                bytes = nil
              end
              io.close if io.respond_to?(:close)
              dest = "#{Downloads}/#{File.basename(entry.name)}"
              File.open(dest, 'w+') { |f| f.write xml }
              Oddb2xml.log "read_xml_from_zip saved as #{dest}"
            end
          end
        end
      else
        Zip::File.foreach(zipfile) do |entry|
          if entry.name =~ target
            Oddb2xml.log "read_xml_from_zip #{__LINE__}: reading #{entry.name}"
            dest = "#{Downloads}/#{File.basename(entry.name)}"
            entry.get_input_stream { |io| xml = io.read }
            File.open(dest, 'w+') { |f| f.write xml }
            Oddb2xml.log "read_xml_from_zip saved as #{dest}"
          end
        end
      end
      xml
    end
  end
  class MigelDownloader < Downloader
    include DownloadMethod
    def download
      @url ||= 'https://github.com/zdavatz/oddb2xml_files/raw/master/NON-Pharma.xls'
      download_as('oddb2xml_files_nonpharma.xls', 'rb')
    end
  end unless SkipMigelDownloader
  class EphaDownloader < Downloader
    include DownloadMethod
    def download
      @url ||= 'https://download.epha.ch/cleaned/matrix.csv'
      file = 'epha_interactions.csv'
      content = download_as(file, 'r')
      FileUtils.rm_f(file, :verbose => true)
      content
    end
  end
  class LppvDownloader < Downloader
    include DownloadMethod
    def download
      @url ||= 'https://raw.githubusercontent.com/zdavatz/oddb2xml_files/master/LPPV.txt'
      download_as('oddb2xml_files_lppv.txt', 'r')
    end
  end
  class ZurroseDownloader < Downloader
    include DownloadMethod
    def download
      @url ||= 'http://pillbox.oddb.org/TRANSFER.ZIP'
      unless @url =~ /^http/
        io = File.open(@url, 'r:iso-8859-1:utf-8')
        content = io.read
        report_download(@url, File.basename(@url))
        Oddb2xml.log("ZurroseDownloader #{__LINE__} download #{@url} @url returns #{content.bytes}")
        content
      else
        file = File.join(Downloads, 'transfer.zip')
        report_download(@url, file)
        unless Oddb2xml.skip_download(file)
          begin
            response = @agent.get(@url)
            response.save_as(file)
            response = nil # win
          rescue Timeout::Error, Errno::ETIMEDOUT
            retrievable? ? retry : raise
          ensure
            Oddb2xml.download_finished(file)
          end
        end
        read_xml_from_zip(/transfer.dat/, file)
        dest = File.join(Downloads, 'transfer.dat')
        File.open(dest, 'r:iso-8859-1:utf-8').read
      end
    end
  end
  class MedregbmDownloader < Downloader
    include DownloadMethod
    def initialize(type=:company)
      @type = type
      case @type
      when :company # betrieb
        action = 'CreateExcelListBetriebs'
      when :person  # medizinalperson
        action = 'CreateExcelListMedizinalPersons'
      else
        action = ''
      end
      url = "https://www.medregbm.admin.ch/Publikation/#{action}"
      super({}, url)
    end
    def download
      file = "medregbm_#{@type.to_s}.txt"
      report_download(@url, file)
      FileUtils.rm_f(file, :verbose => false) # we need it only in the download
    end
  end
  class BagXmlDownloader < Downloader
    def init
      super
      @url ||= 'http://bag.e-mediat.net/SL2007.Web.External/File.axd?file=XMLPublications.zip'
    end
    def download
      file = File.join(WorkDir, 'XMLPublications.zip')
      report_download(@url, file)
      unless Oddb2xml.skip_download(file)
        begin
          response = @agent.get(@url)
          response.save_as(file)
          response = nil # win
        rescue Timeout::Error, Errno::ETIMEDOUT
          retrievable? ? retry : raise
        ensure
          Oddb2xml.download_finished(file)
        end
      end
      content = read_xml_from_zip(/Preparations.xml/, File.join(Downloads, File.basename(file)))
      FileUtils.rm_f(file, :verbose => false) unless defined?(RSpec)
      content
    end
  end
  class RefdataDownloader < Downloader
    def initialize(options={}, type=:pharma)
      @type = (type == :pharma ? 'Pharma' : 'NonPharma')
      url = "http://refdatabase.refdata.ch/Service/Article.asmx?WSDL"
      super(options, url)
    end
    def init
      config = {
        :log_level       => :info,
        :log             => false, # $stdout
        :raise_errors    => true,
        :ssl_version     => :SSLv2,
        :wsdl            => @url
      }
      @client = Savon::Client.new(config)
    end
    def download
      begin
        filename =  "refdata_#{@type}.xml"
        file2save = File.join(Downloads, "refdata_#{@type}.xml")
				soap = %(<?xml version="1.0" encoding="UTF-8"?>
<SOAP-ENV:Envelope xmlns:SOAP-ENV="http://schemas.xmlsoap.org/soap/envelope/" xmlns:ns1="http://refdatabase.refdata.ch/Article_in" xmlns:ns2="http://refdatabase.refdata.ch/">
  <SOAP-ENV:Body>
    <ns2:DownloadArticleInput>
      <ns1:ATYPE>#{@type.upcase}</ns1:ATYPE>
    </ns2:DownloadArticleInput>
  </SOAP-ENV:Body>
  </SOAP-ENV:Envelope>
</ns1:ATYPE></ns2:DownloadArticleInput></SOAP-ENV:Body>
)
        report_download(@url, file2save)
        return IO.read(file2save) if Oddb2xml.skip_download? and File.exists?(file2save)
        FileUtils.rm_f(file2save, :verbose => false)
        response = @client.call(:download, :xml => soap)
        if response.success?
          if xml = response.to_xml
            response = nil # win
            FileUtils.makedirs(Downloads)
            File.open(file2save, 'w+') { |file| file.write xml }
          else
            # received broken data or internal error
            raise StandardError
          end
        else
          raise Timeout::Error
        end
      rescue HTTPI::SSLError
        exit # catch me in Cli class
      rescue Timeout::Error, Errno::ETIMEDOUT
        retrievable? ? retry : raise
      end
      xml
    end
  end
  class SwissmedicDownloader < Downloader
    def initialize(type=:orphan, options = {})
      @type = type
      @options = options
      case @type
      when :orphan
        action = "arzneimittel/00156/00221/00222/00223/00224/00227/00228/index.html?lang=de"
        @xpath = "//div[@id='sprungmarke10_3']//a[@title='Humanarzneimittel']"
      when :package
        action = "arzneimittel/00156/00221/00222/00230/index.html?lang=de"
        @xpath = "//div[@id='sprungmarke10_7']//a[@title='Excel-Version Zugelassene Verpackungen*']"
      end
      url = "https://www.swissmedic.ch/#{action}"
      super(@options, url)
    end
    def download
      @type == file = File.join(Oddb2xml::WorkDir, "swissmedic_#{@type}.xlsx")
      report_download(@url, file)
      if  @options[:calc] and @options[:skip_download] and File.exists?(file) and (Time.now-File.ctime(file)).to_i < 24*60*60
        Oddb2xml.log "SwissmedicDownloader #{__LINE__}: Skip downloading #{file} #{File.size(file)} bytes"
        return File.expand_path(file)
      end
      begin
        FileUtils.rm(File.expand_path(file), :verbose => !defined?(RSpec)) if File.exists?(File.expand_path(file))
        page = @agent.get(@url)
        if !page.class.is_a?(String) and link_node = page.search(@xpath).first
          link = Mechanize::Page::Link.new(link_node, @agent, page)
          response = link.click
          response.save_as(file)
          response = nil # win
        end
        return File.expand_path(file)
      rescue Timeout::Error, Errno::ETIMEDOUT
        retrievable? ? retry : raise
      ensure
        Oddb2xml.download_finished(file, false)
      end
      return File.expand_path(file)
    end
  end
  class SwissmedicInfoDownloader < Downloader
    def init
      super
      @agent.ignore_bad_chunking = true
      @url ||= "http://download.swissmedicinfo.ch/Accept.aspx?ReturnUrl=%2f"
    end
    def download
      file = File.join(Downloads, "swissmedic_info.zip")
      report_download(@url, file)
      FileUtils.rm_f(file, :verbose => false) unless Oddb2xml.skip_download?
      begin
        response = nil
        if home = @agent.get(@url)
          form = home.form_with(:id => 'Form1')
          bttn = form.button_with(:name => 'ctl00$MainContent$btnOK')
          if page = form.submit(bttn)
            form = page.form_with(:id => 'Form1')
            bttn = form.button_with(:name => 'ctl00$MainContent$BtnYes')
            response = form.submit(bttn)
          end
        end
        if response
          response.save_as(file)
          response = nil # win
        end
      rescue Timeout::Error, Errno::ETIMEDOUT
        retrievable? ? retry : raise
      rescue NoMethodError
        # pass
      ensure
        Oddb2xml.download_finished(file)
      end unless File.exists?(file)
      read_xml_from_zip(/^AipsDownload_/iu, file)
    end
  end
end
